#**Alexandros Skondras AM: f3352119**
**Page Ranking - Numerical Optimization and Large Scale Linear Algebra**

In the following code, apart from essential cells, there are also some that contribute to making ascertainments, drawing conclusions and indicate my train of thought.

In [1]:
import os
import numpy as np
import pandas as pd
from scipy.sparse import csr_matrix
from scipy import stats

In [2]:
cwd = os.getcwd()
cwd

'/content'

In [3]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [4]:
os.chdir('/content/drive/MyDrive')          # I have already mounted my drive

In [5]:
cwd = os.getcwd()                           # just to check
cwd

'/content/drive/MyDrive'

In [6]:
# reading the file and checking its length
with open("out.web-BerkStan") as initial_file:
  file = initial_file.read()

len(file)

107347903

**An approach**
just to see what I am dealing with.

In [ ]:
print(file[:42])

% asym unweighted
% 7600595 680486 617094



In [ ]:
print(file[:43])

% asym unweighted
% 7600595 680486 617094
1


In [ ]:
print(file[:52])

% asym unweighted
% 7600595 680486 617094
1 2 
1 3 



In [ ]:
file2 = file[42:]
print(file2[:9])

1 2 
1 3 


**Preprocessing:**

In [ ]:
# splitting the string file and checking where the inlinks and outlinks columns start
file1 = file.split()
print(len(file1))
print(file1[-1:])
print(file1[:7])
file1 = file1[7:]
print(file1[:4])

15201197
['684761']
['%', 'asym', 'unweighted', '%', '7600595', '680486', '617094']
['1', '2', '1', '3']


In [ ]:
data = pd.DataFrame(list(zip([int(i)-1 for i in file1[::2]], [int(i)-1 for i in file1[1::2]])),
               columns = ['Starting_node', 'Ending_node'])

In [ ]:
data

Starting_node  Ending_node
0                    0            1
1                    0            2
2                    0            3
3                    0            4
4                    0            5
...                ...          ...
7600590         684759       684750
7600591         684759       684756
7600592         684759       684757
7600593         684759       684758
7600594         684759       684760

[7600595 rows x 2 columns]

In [ ]:
data = data.sort_values("Starting_node", ascending=True)   # sorting the dataframe rows in ascending order of Startin_node column
                                                           # because the method np.unique() does NOT keep preserve the order of the Startin_node column
                                                           # when calculating the frequencies of its elements through return_counts

In [ ]:
data

Starting_node  Ending_node
0                    0            1
1                    0            2
2                    0            3
3                    0            4
4                    0            5
...                ...          ...
7600537         685229       650231
7600538         685229       650232
7600539         685229       650236
7600540         685229       650238
7600541         685229       650241

[7600595 rows x 2 columns]

In [ ]:
unique_elements, frequencies = np.unique(data.Starting_node, return_counts=True)

In [ ]:
frequencies                                       # frequency of each starting node(page) // original order is NOT preserved --> 9 0s, 8 1s, 10 2s, ...

array([ 9,  8, 10, ...,  2,  8,  8])

In [ ]:
probs_list_to_add_to_data = [[1/i]*i for i in frequencies]        # list of lists that contain the probs of starting page-->its links // sequence is NOT preserved

In [ ]:
probs_list_to_add_to_data

[[0.1111111111111111,
  0.1111111111111111,
  0.1111111111111111,
  0.1111111111111111,
  0.1111111111111111,
  0.1111111111111111,
  0.1111111111111111,
  0.1111111111111111,
  0.1111111111111111],
 [0.125, 0.125, 0.125, 0.125, 0.125, 0.125, 0.125, 0.125],
 [0.1, 0.1, 0.1, 0.1, 0.1, 0.1, 0.1, 0.1, 0.1, 0.1],
 [0.058823529411764705,
  0.058823529411764705,
  0.058823529411764705,
  0.058823529411764705,
  0.058823529411764705,
  0.058823529411764705,
  0.058823529411764705,
  0.058823529411764705,
  0.058823529411764705,
  0.058823529411764705,
  0.058823529411764705,
  0.058823529411764705,
  0.058823529411764705,
  0.058823529411764705,
  0.058823529411764705,
  0.058823529411764705,
  0.058823529411764705],
 [0.08333333333333333,
  0.08333333333333333,
  0.08333333333333333,
  0.08333333333333333,
  0.08333333333333333,
  0.08333333333333333,
  0.08333333333333333,
  0.08333333333333333,
  0.08333333333333333,
  0.08333333333333333,
  0.08333333333333333,
  0.08333333333333333],
 [0

In [ ]:
# flattening the list probs_list_to_add_to_data
probabilities_flat_list = []
for i in range(len(probs_list_to_add_to_data)):
  for j in probs_list_to_add_to_data[i]:
    probabilities_flat_list.append(j)

In [ ]:
len(np.asarray(data.Starting_node))               # number of connections (links)

7600595

In [ ]:
len(probabilities_flat_list)                      # it coincides correctly with the cell above

7600595

In [ ]:
probabilities_flat_list                           # probabilities' list

[0.1111111111111111,
 0.1111111111111111,
 0.1111111111111111,
 0.1111111111111111,
 0.1111111111111111,
 0.1111111111111111,
 0.1111111111111111,
 0.1111111111111111,
 0.1111111111111111,
 0.125,
 0.125,
 0.125,
 0.125,
 0.125,
 0.125,
 0.125,
 0.125,
 0.1,
 0.1,
 0.1,
 0.1,
 0.1,
 0.1,
 0.1,
 0.1,
 0.1,
 0.1,
 0.058823529411764705,
 0.058823529411764705,
 0.058823529411764705,
 0.058823529411764705,
 0.058823529411764705,
 0.058823529411764705,
 0.058823529411764705,
 0.058823529411764705,
 0.058823529411764705,
 0.058823529411764705,
 0.058823529411764705,
 0.058823529411764705,
 0.058823529411764705,
 0.058823529411764705,
 0.058823529411764705,
 0.058823529411764705,
 0.058823529411764705,
 0.08333333333333333,
 0.08333333333333333,
 0.08333333333333333,
 0.08333333333333333,
 0.08333333333333333,
 0.08333333333333333,
 0.08333333333333333,
 0.08333333333333333,
 0.08333333333333333,
 0.08333333333333333,
 0.08333333333333333,
 0.08333333333333333,
 0.09090909090909091,
 0.0909090

Notice that earlier I subtracted 1 to each node(page) number while making the dataframe. If I hadn't done that, the sparse P would have to have shape of (685231, 685231). This is because the enumeration of the pages starts from 1, but python matrices dimensions start from dimension zero, thus an empty row 0 and a column 0 would also have been included in the sparse matrix.

**So, now the sparse matrix is of dimensions 685230x685230 and page 0 is actually initial page 1, page 1 is actually initial page 2 and so on.**

In [ ]:
Psparse = csr_matrix((probabilities_flat_list, (np.asarray(data.Starting_node), np.asarray(data.Ending_node))), shape=(max(data.Starting_node)+1, max(data.Starting_node)+1)) # max+1=685230

In [ ]:
print(Psparse)

  (0, 1)	0.1111111111111111
  (0, 2)	0.1111111111111111
  (0, 3)	0.1111111111111111
  (0, 4)	0.1111111111111111
  (0, 5)	0.1111111111111111
  (0, 6)	0.1111111111111111
  (0, 7)	0.1111111111111111
  (0, 8)	0.1111111111111111
  (0, 9)	0.1111111111111111
  (1, 3)	0.125
  (1, 5)	0.125
  (1, 6)	0.125
  (1, 7)	0.125
  (1, 8)	0.125
  (1, 9)	0.125
  (1, 105)	0.125
  (1, 106)	0.125
  (2, 1)	0.1
  (2, 3)	0.1
  (2, 4)	0.1
  (2, 5)	0.1
  (2, 6)	0.1
  (2, 7)	0.1
  (2, 8)	0.1
  (2, 9)	0.1
  :	:
  (685225, 685220)	0.14285714285714285
  (685225, 685221)	0.14285714285714285
  (685225, 685222)	0.14285714285714285
  (685225, 685223)	0.14285714285714285
  (685225, 685224)	0.14285714285714285
  (685226, 685197)	0.5
  (685226, 685198)	0.5
  (685227, 685228)	0.5
  (685227, 685229)	0.5
  (685228, 9)	0.125
  (685228, 650214)	0.125
  (685228, 650230)	0.125
  (685228, 650231)	0.125
  (685228, 650232)	0.125
  (685228, 650236)	0.125
  (685228, 650238)	0.125
  (685228, 650241)	0.125
  (685229, 9)	0.125
  (685229, 6

**Now I am going to check if I created the sparse matrix correctly:**<br>More particulary, I am going to check a random instance-page --> e.g. if the connections of page 685229 (initially page 685230 and page with the highest number) coincide with the ones in the dataframe I created.

In [ ]:
Psparse

<685230x685230 sparse matrix of type '<class 'numpy.float64'>'
	with 7600595 stored elements in Compressed Sparse Row format>

In [ ]:
print(data.Starting_node.value_counts()[685229], data.Ending_node.value_counts()[685229])

8 1


In [ ]:
print(Psparse[(685229,)])               # notice there are 8 links here

  (0, 9)	0.125
  (0, 650214)	0.125
  (0, 650230)	0.125
  (0, 650231)	0.125
  (0, 650232)	0.125
  (0, 650236)	0.125
  (0, 650238)	0.125
  (0, 650241)	0.125


In [ ]:
data.Starting_node.loc[data.Ending_node==685229]

7600525    685227
Name: Starting_node, dtype: int64

In [ ]:
print(Psparse[(685227,)])               # notice the link (685227, 685229) exists so in total there are correctly 8+1 links from or to 685230

  (0, 685228)	0.5
  (0, 685229)	0.5


In [ ]:
def power_method(a, tol, dimension):
  # a = 0.85
  # tol = 10**(-8)                                 # tolerance
  alpha = np.ones(max(dimension)+1)                # max+1=685230
  for i in unique_elements:
    alpha[i] = 0
  n = max(dimension)+1
  v = np.repeat(1/n, n)
  previous_x = np.repeat(1/n, n)

  iter = 0
  condition = 1
  while condition>tol:
    new_x = a*previous_x.T*Psparse + (a*previous_x.T.dot(alpha) + (1-a))*v.T
    condition = np.linalg.norm(new_x-previous_x)
    iter += 1
    previous_x = new_x
  
  return sum(previous_x), iter, new_x     # sum of normalized "pageranks", converges in __ iterations, pageranks' vector

**Case α=0.85:**

In [ ]:
question_1 = power_method(0.85, 10**(-8), data.Starting_node)       # ~3 seconds
question_1

(0.9999999999988669,
 71,
 array([1.94623518e-05, 4.35093007e-06, 2.93325615e-06, ...,
        2.23198585e-07, 3.18057983e-07, 3.18057983e-07]))

In [ ]:
top_100_085 = np.flip(np.argsort(question_1[2])[-100:])

In [ ]:
# question_1[2].argsort()[-100:][::-1]

In [ ]:
least_important_100_085 = np.argsort(question_1[2])[:100]           # I will use this later for Question (3)

**Case α=0.99:**

In [ ]:
question_2 = power_method(0.99, 10**(-8), data.Starting_node)       # ~35 seconds
question_2

(1.0000000000001101,
 1216,
 array([8.08614919e-06, 2.22442457e-06, 1.41575252e-06, ...,
        1.66072208e-08, 2.48277951e-08, 2.48277951e-08]))

In [ ]:
top_100_099 = np.flip(np.argsort(question_2[2])[-100:])

**Comparison of the two cases:**

In [ ]:
print("Number of pages that have the same rank in cases α=0.5 and α=0.99:", np.sum(top_100_085==top_100_099))

Number of pages that have the same rank in cases α=0.5 and α=0.99: 1


Only 1 page out of the 100 most important ones kept its exact rank when variable α changed from 0.85 to 0.99.

In [ ]:
count_common_pages = 0                         # count how many common pages there are in the top 100 pages
count_similar_rank_pages = 0                   # count how many pages there are with similar rank
for i in range(len(top_100_085)):
  if top_100_085[i] in top_100_099:
    count_common_pages += 1
    if abs(top_100_099.tolist().index(top_100_085[i])-i)<=5:
      count_similar_rank_pages += 1

print("Pages that remain in the top 100 pages with the highest ranks:", count_common_pages)
print("Pages with similar rank:", count_similar_rank_pages)

Pages that remain in the top 100 pages with the highest ranks: 65
Pages with similar rank: 28


Notice that with a change of α from 0.85 to 0.99, 35 pages of the first top 100 (most important) do not remain in the first top 100 after the change is implemented.<br>In addition to that, from the 65 ones that remain in the top 100 pages, only 28 of them have a similar rank, +-5 indices from their initial rank.<br>The speed of convergence in the case of α=0.99 is also significantly slower.

**Now, I am going to define a new function for the power_method but also with max(iterations) as input, in order to timesnap the convergence of the values of vector π.**

In [ ]:
def power_method_with_max_iter(a, tol, dimension, max_iterations):

  alpha = np.ones(max(dimension)+1)                # max+1=685230
  for i in unique_elements:
    alpha[i] = 0
  n = max(dimension)+1
  v = np.repeat(1/n, n)
  previous_x = np.repeat(1/n, n)

  iter = 0
  condition = 1
  while condition>tol and iter<max_iterations:                         # !!!!! Here is the only change added to the function
    new_x = a*previous_x.T*Psparse + (a*previous_x.T.dot(alpha) + (1-a))*v.T
    condition = np.linalg.norm(new_x-previous_x)
    iter += 1
    previous_x = new_x
  
  return sum(previous_x), iter, new_x              # sum of normalized "pageranks", converges in __ iterations, pageranks' vector

Remember that with α=0.85 and tol=10e-8, 71 iterations take place before convergence.

In [ ]:
iter_snaps = [15, 30, 45, 50, 65, 70, 71]
pi_vectors_list = []                               # here, the π vectors will be added of 15,30,45,60,70,71 iterations
for i in iter_snaps:
  pi_vectors_list.append(power_method_with_max_iter(0.85, 10**(-8), data.Starting_node, i)[2])

So, now I need to compare all the values of each snap(at a specific iteration point) π vector with the totally converged values of the π vector with max_iterations=71.<br>I will use values 1e-11, 1e-12, 1e-13 as conditions to determine whether a node(page) has converged or not.

In [ ]:
for j in range(len(pi_vectors_list)):              # all iter_snaps
  count_of_converged_values = 0
  for i in range(len(pi_vectors_list[0])):
    if abs(pi_vectors_list[j][i]-pi_vectors_list[-1][i])<=10**(-11):
      count_of_converged_values += 1

  print("Nodes that still haven't converged after iteration", iter_snaps[j], ":", max(data.Starting_node)+1-count_of_converged_values)
                                                                                # max(data.Starting_node)+1 = 685230 -->total pages

Nodes that still haven't converged after iteration 15 : 681201
Nodes that still haven't converged after iteration 30 : 336786
Nodes that still haven't converged after iteration 45 : 132098
Nodes that still haven't converged after iteration 50 : 84340
Nodes that still haven't converged after iteration 65 : 9948
Nodes that still haven't converged after iteration 70 : 7113
Nodes that still haven't converged after iteration 71 : 0


In [ ]:
for j in range(len(pi_vectors_list)):          # all iter_snaps
  count_of_converged_values = 0
  for i in range(len(pi_vectors_list[0])):
    if abs(pi_vectors_list[j][i]-pi_vectors_list[-1][i])<=10**(-12):
      count_of_converged_values += 1
      
  print("Nodes that still haven't converged after iteration", iter_snaps[j], ":", max(data.Starting_node)+1-count_of_converged_values)
                                                                                # max(data.Starting_node)+1 = 685230 -->total pages

Nodes that still haven't converged after iteration 15 : 684881
Nodes that still haven't converged after iteration 30 : 676518
Nodes that still haven't converged after iteration 45 : 286017
Nodes that still haven't converged after iteration 50 : 233501
Nodes that still haven't converged after iteration 65 : 53704
Nodes that still haven't converged after iteration 70 : 24976
Nodes that still haven't converged after iteration 71 : 0


In [ ]:
for j in range(len(pi_vectors_list)):          # all iter_snaps
  count_of_converged_values = 0
  for i in range(len(pi_vectors_list[0])):
    if abs(pi_vectors_list[j][i]-pi_vectors_list[-1][i])<=10**(-13):
      count_of_converged_values += 1
      
  print("Nodes that still haven't converged after iteration", iter_snaps[j], ":", max(data.Starting_node)+1-count_of_converged_values)
                                                                                # max(data.Starting_node)+1 = 685230 -->total pages

Nodes that still haven't converged after iteration 15 : 685204
Nodes that still haven't converged after iteration 30 : 683913
Nodes that still haven't converged after iteration 45 : 512318
Nodes that still haven't converged after iteration 50 : 371067
Nodes that still haven't converged after iteration 65 : 178392
Nodes that still haven't converged after iteration 70 : 63275
Nodes that still haven't converged after iteration 71 : 0


**As indicated in all 3 iteration point results above, it is clear that the speed of convergence is not the same for every node.**<br>Now, I'll compare the speed of convergence of the top 100 most important pages to the speed of convergence of the 100 least important ones.

In [ ]:
top_100_list_of_lists = []                       # contains 7 lists (one per iter_snap) with the values of the 100 most important nodes
least_important_list_of_lists = []               # contains 7 lists (one per iter_snap) with the values of the 100 least important nodes
for pi_vector in pi_vectors_list:
  a_list = [pi_vector[i] for i in top_100_085]
  top_100_list_of_lists.append(a_list)
  b_list = [pi_vector[i] for i in least_important_100_085]
  least_important_list_of_lists.append(b_list)

In [ ]:
for i in range(len(top_100_list_of_lists)):
  count_important_converged_values = 0
  count_least_important_converged_values = 0
  for j in range(len(top_100_list_of_lists[0])):
    if abs(top_100_list_of_lists[i][j]-top_100_list_of_lists[6][j])<=10**(-9):
      count_important_converged_values += 1
    if abs(least_important_list_of_lists[i][j]-least_important_list_of_lists[6][j])<=10**(-9):
      count_least_important_converged_values += 1

  print("Iteration", iter_snaps[i], "| Important nodes not converged:", 100-count_important_converged_values, "| Least important nodes not converged:", 100-count_least_important_converged_values)

Iteration 15 | Important nodes not converged: 100 | Least important nodes not converged: 0
Iteration 30 | Important nodes not converged: 100 | Least important nodes not converged: 0
Iteration 45 | Important nodes not converged: 99 | Least important nodes not converged: 0
Iteration 50 | Important nodes not converged: 94 | Least important nodes not converged: 0
Iteration 65 | Important nodes not converged: 10 | Least important nodes not converged: 0
Iteration 70 | Important nodes not converged: 0 | Least important nodes not converged: 0
Iteration 71 | Important nodes not converged: 0 | Least important nodes not converged: 0


In [ ]:
for i in range(len(top_100_list_of_lists)):
  count_important_converged_values = 0
  count_least_important_converged_values = 0
  for j in range(len(top_100_list_of_lists[0])):
    if abs(top_100_list_of_lists[i][j]-top_100_list_of_lists[6][j])<=10**(-10):
      count_important_converged_values += 1
    if abs(least_important_list_of_lists[i][j]-least_important_list_of_lists[6][j])<=10**(-10):
      count_least_important_converged_values += 1

  print("Iteration", iter_snaps[i], "| Important nodes not converged:", 100-count_important_converged_values, "| Least important nodes not converged:", 100-count_least_important_converged_values)

Iteration 15 | Important nodes not converged: 100 | Least important nodes not converged: 0
Iteration 30 | Important nodes not converged: 100 | Least important nodes not converged: 0
Iteration 45 | Important nodes not converged: 100 | Least important nodes not converged: 0
Iteration 50 | Important nodes not converged: 100 | Least important nodes not converged: 0
Iteration 65 | Important nodes not converged: 69 | Least important nodes not converged: 0
Iteration 70 | Important nodes not converged: 16 | Least important nodes not converged: 0
Iteration 71 | Important nodes not converged: 0 | Least important nodes not converged: 0


In [ ]:
for i in range(len(top_100_list_of_lists)):
  count_important_converged_values = 0
  count_least_important_converged_values = 0
  for j in range(len(top_100_list_of_lists[0])):
    if abs(top_100_list_of_lists[i][j]-top_100_list_of_lists[6][j])<=10**(-11):
      count_important_converged_values += 1
    if abs(least_important_list_of_lists[i][j]-least_important_list_of_lists[6][j])<=10**(-11):
      count_least_important_converged_values += 1

  print("Iteration", iter_snaps[i], "| Important nodes not converged:", 100-count_important_converged_values, "| Least important nodes not converged:", 100-count_least_important_converged_values)

Iteration 15 | Important nodes not converged: 100 | Least important nodes not converged: 100
Iteration 30 | Important nodes not converged: 100 | Least important nodes not converged: 0
Iteration 45 | Important nodes not converged: 100 | Least important nodes not converged: 0
Iteration 50 | Important nodes not converged: 100 | Least important nodes not converged: 0
Iteration 65 | Important nodes not converged: 99 | Least important nodes not converged: 0
Iteration 70 | Important nodes not converged: 69 | Least important nodes not converged: 0
Iteration 71 | Important nodes not converged: 0 | Least important nodes not converged: 0


Using 1e-09, 1e-10, 1e-11 as conditions to determine whether a node(page) has converged or not, **it is clear from the results above that the least important nodes converge much faster to their final values than the most important nodes.**

Now, I will choose the 10 most important nodes and 500 least important ones and I will add node number 685230, which will point to the 10 important ones and will be pointed by the 500 least important ones.

In [ ]:
top_10_085 = np.flip(np.argsort(question_1[2])[-10:])
least_important_500_085 = np.argsort(question_1[2])[:500]

In [ ]:
df1 = {'Starting_node':max(data.Starting_node)+1, 'Ending_node':top_10_085}
df_1 = pd.DataFrame(df1)
df2 = {'Starting_node':least_important_500_085, 'Ending_node':max(data.Starting_node)+1}
df_2 = pd.DataFrame(df2)
data_1 = pd.concat([data, df_1, df_2], axis = 0)

I follow the same preprocessing now to create the sparse matrix.

In [ ]:
data_1 = data_1.sort_values("Starting_node", ascending=True)         # sorting the df data_1 by column Starting_node in ascending order

unique_elements, frequencies = np.unique(data_1.Starting_node, return_counts=True)
probs_list_to_add_to_data = [[1/i]*i for i in frequencies]

# flattening the list probs_list_to_add_to_data
probabilities_flat_list = []
for i in range(len(probs_list_to_add_to_data)):
  for j in probs_list_to_add_to_data[i]:
    probabilities_flat_list.append(j)

In [ ]:
Psparse4 = csr_matrix((probabilities_flat_list, (np.asarray(data_1.Starting_node), np.asarray(data_1.Ending_node))), shape=(max(data_1.Starting_node)+1, max(data_1.Starting_node)+1)) # max+1=685231

Redefining the power_method adding to it the P_sparse matrix as input:

In [ ]:
def power_method_P_sparse(a, tol, dimension, P_sparse):

  alpha = np.ones(max(dimension)+1)                # max+1=685231
  for i in unique_elements:
    alpha[i] = 0
  n = max(dimension)+1
  v = np.repeat(1/n, n)
  previous_x = np.repeat(1/n, n)

  iter = 0
  condition = 1
  while condition>tol:
    new_x = a*previous_x.T*P_sparse + (a*previous_x.T.dot(alpha) + (1-a))*v.T
    condition = np.linalg.norm(new_x-previous_x)
    iter += 1
    previous_x = new_x
  
  return sum(previous_x), iter, new_x     # sum of normalized "pageranks", converges in __ iterations, pageranks' vector

In [ ]:
question_4 = power_method_P_sparse(0.85, 10**(-8), data_1.Starting_node, Psparse4)      # ~3 seconds
print("Using the power method the new node, 685230, got the rank:", len(question_4[2])-stats.rankdata(question_4[2], method='ordinal')[-1] + 1)

Using the power method the new node, 685230, got the rank: 2925


So, despite the fact that the 500 least important nodes point to the new node, the 10 most important nodes that get pointed by the new one, pull the new node to a **relatively significantly high rank** compared to the total numbers of nodes(685231).

This time I will add a node to the initial nodes, which gets pointed by the 10 most important nodes and points to the 10 least important ones.

In [ ]:
least_important_10_085 = np.argsort(question_1[2])[:10]
# I already have the 10 most important nodes

I am going to follow the same preprocessing:

In [ ]:
df1 = {'Starting_node':top_10_085, 'Ending_node':max(data.Starting_node)+1}
df_1 = pd.DataFrame(df1)
df2 = {'Starting_node':max(data.Starting_node)+1, 'Ending_node':least_important_10_085}
df_2 = pd.DataFrame(df2)
data_2 = pd.concat([data, df_1, df_2], axis = 0)

In [ ]:
data_2 = data_2.sort_values("Starting_node", ascending=True)         # sorting the df data_1 by column Starting_node in ascending order

unique_elements, frequencies = np.unique(data_2.Starting_node, return_counts=True)
probs_list_to_add_to_data = [[1/i]*i for i in frequencies]

# flattening the list probs_list_to_add_to_data
probabilities_flat_list = []
for i in range(len(probs_list_to_add_to_data)):
  for j in probs_list_to_add_to_data[i]:
    probabilities_flat_list.append(j)

In [ ]:
Psparse5 = csr_matrix((probabilities_flat_list, (np.asarray(data_2.Starting_node), np.asarray(data_2.Ending_node))), shape=(max(data_2.Starting_node)+1, max(data_2.Starting_node)+1)) # max+1=685231

In [ ]:
question_5 = power_method_P_sparse(0.85, 10**(-8), data_2.Starting_node, Psparse5)      # ~3 seconds
print("Using the power method the new node, 685230, got the rank:", len(question_5[2])-stats.rankdata(question_5[2], method='ordinal')[-1] + 1)
# Remember! The new node is numbered 685230 because the initial ones range from 0-286229
# rank 1 is the best rank

Using the power method the new node, 685230, got the rank: 11


It is evident that, this time, the new node got an even higher rank than before. So, it seems that being pointed by important nodes-pages and pointing to unimportant pages is "stronger" than pointing to important ones and being pointed by unimportant ones. The more a page gets recommended-pointed by other important pages, the higher the importance of the node-page gets.

In [ ]:
ranks = len(question_5[2])-stats.rankdata(question_5[2], method='ordinal') + 1       # node 0 has a rank of ranks[0], node 1-->rank ranks[1], ...
print("So, from being in the top 10 most important nodes:")
for node in top_10_085:
  print("Node", node, "now is given the rank", str(ranks[node])+".")

So, from being in the top 10 most important nodes:
Node 288238 now is given the rank 1.
Node 9 now is given the rank 2.
Node 225465 now is given the rank 3.
Node 54130 now is given the rank 4.
Node 1283 now is given the rank 5.
Node 564114 now is given the rank 6.
Node 288346 now is given the rank 7.
Node 65122 now is given the rank 10.
Node 113 now is given the rank 8.
Node 114 now is given the rank 9.


The aforementioned results indicate that even though they all point to a new node, their rank does not get affected (not significantly at least-notice nodes 113 and 114 surpassed node 65122). They remain in the top 10 most important nodes.

In [ ]:
ranks = len(question_5[2])-stats.rankdata(question_5[2], method='ordinal') + 1       # node 0 has a rank of ranks[0], node 1-->rank ranks[1], ...
print("So, from being in the 10 least important nodes:")
for node in least_important_10_085:
  print("Node", node, "now is given the rank", str(ranks[node])+".")

So, from being in the 10 least important nodes:
Node 179980 now is given the rank 367.
Node 648795 now is given the rank 365.
Node 648796 now is given the rank 364.
Node 648797 now is given the rank 363.
Node 648798 now is given the rank 362.
Node 648799 now is given the rank 361.
Node 648800 now is given the rank 360.
Node 648801 now is given the rank 359.
Node 648802 now is given the rank 358.
Node 476329 now is given the rank 366.


It is apparent that the 10 least important nodes-pages have moved up ranks greatly. This is because they get pointed by the new node, which gets pointed-recommended by the 10 most important nodes. So, the most important nodes practically pull the least important ones towards the top when they point at them directly or indirectly (i.e. in this case through the new node).

Note that the 10 least important nodes have different ranks but in fact have the same value score in the produced vector π by the power method. It is the function of rankdata, which assigns a different but consecutive rank to each among them, although their π scores are the same.